OpenSim/Simulation/Test/testModelInterface.cpp

```diff
@@ -0,0 +1,150 @@
+        ASSERT(!model.isObjectUpToDateWithProperties());
+
+        model.finalizeFromProperties();
+        ASSERT(model.isObjectUpToDateWithProperties());
+
+        // get writable access to another Set for the purpose of editing
+        auto& bodies = model.updBodySet();
+        // for example, bodies[1].upd_mass() = 0.05;
+        ASSERT(!model.isObjectUpToDateWithProperties());
+
+        model.finalizeFromProperties();
+        ASSERT(model.isObjectUpToDateWithProperties());
+
+        // make an edit through model's ComponentList access
+        for (auto& body : model.updComponentList<Body>()) {
+            body.upd_mass_center() = SimTK::Vec3(0);
+            break;
+        }
+        ASSERT(!model.isObjectUpToDateWithProperties());
+
+        SimTK::State dummy_state;
+
+        ASSERT_THROW(ComponentHasNoSystem, model.getSystem());
+        ASSERT_THROW(ComponentHasNoSystem, model.realizeDynamics(dummy_state));
+        ASSERT_THROW(ComponentHasNoSystem, 
+                     model.computeStateVariableDerivatives(dummy_state));
+        // should not be able to create a Manager either
+        ASSERT_THROW(ComponentHasNoSystem,
+                     Manager manager(model));
+
+        // get a valid System and corresponding state
+        SimTK::State state = model.initSystem();
+        Manager manager(model);
+        state.setTime(0.0);
+
+        // this should invalidate the System
+        model.finalizeFromProperties();
+
+        // verify that finalizeFromProperties() wipes out the underlying System
+        ASSERT_THROW(ComponentHasNoSystem, model.getSystem());
+
+        // should not be able to "trick" the manager into integrating a model
+        // given a stale but compatible state
+        ASSERT_THROW(ComponentHasNoSystem, manager.integrate(state, 1.));
+
+        // once again, get a valid System and corresponding state
+        state = model.initSystem();
+        
+        // Test for the effects of calling finalizeConnections() on the model
+        // after initSystem() has been called.
+        // In this case, there are no changes to the connections to be finalized.
+        model.finalizeConnections(model);
+
+        // verify that finalizeConnections() does not wipe out the underlying 
+        // System when there are no changes to the connections
+        auto& sys = model.getSystem();
+
+        auto elbowInHumerus = new PhysicalOffsetFrame("elbow_in_humerus",
+            model.getComponent<Body>("r_humerus"),
+            SimTK::Transform(SimTK::Vec3(0, -0.33, 0.0)) );
+
+        model.addComponent(elbowInHumerus);
+
+        // establish the new offset frame as part of the model but not
+        // used by any joints, constraints or forces
+        state = model.initSystem();
+
+        // update the elbow Joint and connect its socket to the new frame
+        Joint& elbow = model.updComponent<Joint>("r_elbow");
+        elbow.connectSocket_parent_frame(*elbowInHumerus);
```

Comment:
Consider documenting that this call requires recreating the system. Places where this could be added:

- https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ComponentSocket.h#L912 and similar functions in other macros.
- https://github.com/opensim-org/opensim-core/blob/master/OpenSim/Common/ComponentSocket.h#L142 and similar `connect()` functions in derived classes.

Would be good to add a similar check for `connectInput_()`; at the very least, we should add a TODO for that.